### PR TITLE
Hash code references

### PR DESCRIPTION
**Pull request by namanjain**
This change integrates the code references logic in Bionic's caching
strategy. It gets us a step closer to having a robust automatic caching
strategy.

Here is the list of things left after this change:
- hashing classes
- documentation
- caching individual object hashes
- investigate if we can hash module or package versions
- verify that we hash all Python constant types
- add more tests, maybe crowdsource ideas for more comprehensive tests
from the rest of the team
- version.suppress_bytecode_warnings TODO
- any other TODOs in code hasher and code reference logic

### CODE REVIEW

**Review by jqmp — COMMENTED** [review]
It's really exciting to see this coming together! A few thoughts:

1. Would it make sense to just combine this with your other PR?
2. We still need to add release notes and update the documentation, right?
3. Have you noticed any performance impact from these changes? Do the tests take longer?
4. These changes affect assisted versioning too, right? Do we need to change how versioning errors are reported?
5. The tests here are good, but shouldn't we also have some tests on more complex code? Your hashing algorithm is quite complex and stateful, so it feels like it has a lot of state space that isn't explored by these relatively small tests.

That last point is a hard one. Some dumb ideas:
1. Using the `ml_workflow_cli` example. (It's not actually that complex, though -- and we wouldn't want to make our tutorial code more complex for the sake of this test. But it would actually be great for our examples to be run as part of our test suite.)
2. Generating random code. (You'll only be able to test the constructs that you think to generate, but at least you'll be able to test how the different combinations interact.)
3. Importing (or vendoring) some open source library and trying to hash some functions in it; then monkey-patching some of the other functions or constants. (Kind of hacky but saves you from coming up with your own code.)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -18,7 +18,13 @@ repos:
     hooks:
     -   id: trailing-whitespace
     -   id: end-of-file-fixer
-    -   id: check-yaml
+    # TODO: This throws an error on MacOS Big Sur. Updating the rev
```
[the remark this answers is not in the document]
Comment (namanjain) [reply .pre-commit-config.yaml:21]: I noticed this problem after updating my OS. I couldn't find a related issue online, which is why I think it is related to Big Sur and `check-yaml` hook.

File: bionic/decorators.py [hunk 2]
```diff
@@ -59,15 +75,37 @@ def version(major=None, minor=None):
...
+    suppress_bytecode_warnings: Boolean (default True)
+        Whether warnings from this entity's bytecode analysis should be ignored.
+
     Returns
     -------
     Function:
         A decorator which can be applied to an entity function.
     """
 
+    if ignore_bytecode is not None and not isinstance(ignore_bytecode, bool):
```
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/decorators.py:90]: Because that's what the other fields in version decorator are doing as well. I can move it here though.

File: bionic/code_hasher.py [hunk 3]
```diff
@@ -43,14 +45,15 @@ class CodeHasher:
     it doesn't actually achieve the ideal behavior we described above.
```
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:45]: Nope. Good catch!

File: bionic/code_hasher.py [hunk 4]
```diff
@@ -144,41 +147,74 @@ def _ingest(self, obj):
...
-                """
-            )
-            warnings.warn(message)
-
-    def _ingest_code(self, code):
-        # TODO: Find references for the code and analyze references.
+            if not self._suppress_warnings:
+                message = oneline(
+                    f"""
+                    Found a complex object {obj!r} of type {type(obj)!r}
```
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:201]: We can't tell the line number with the current code. We'll have to track where the objects are coming from, but that shouldn't be too hard I think. I think we can get variable name from `obj.__name__` though. I'll leave a TODO for this too.

File: bionic/datatypes.py [hunk 5]
```diff
@@ -302,6 +305,10 @@ class CodeVersion:
 
     major: str = attr.ib(converter=str_from_version_value)
     minor: str = attr.ib(converter=str_from_version_value)
+    ignore_bytecode: bool = attr.ib(converter=attr.converters.default_if_none(False))
```
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/datatypes.py:308]: They aren't part of the "version" per se, but they are ingested into Bionic from `@version` decorator that directly translates to `CodeVersion`.

Also, even though `suppress_bytecode_warnings` doesn't define the version, `ignore_bytecode` does. For automatic version, the provenance hash of a function with version `1.0` will be different when `ignore_bytecode` is `True` vs `False`.

File: bionic/code_hasher.py [hunk 6]
```diff
@@ -66,7 +69,7 @@ def _ingest_raw_prefix_and_bytes(self, type_prefix, obj_bytes=b""):
         self._hash.update(PREFIX_SEPARATOR)
         self._hash.update(obj_bytes)
 
-    def _check_and_ingest(self, obj):
+    def _check_and_ingest(self, obj, code_context=None):
```
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:72]: `code_context` will always be `None` for the first call. Do you think it's easier to read if we pass a `DEFAULT_CONTEXT` instead?

File: tests/test_flow/test_persistence_random.py [hunk 7]
```diff
@@ -200,7 +200,7 @@ def _add_binding_to_flow(self, binding):
             output_value_fragment = binding.value_code_fragment_for_outputs()
 
         vars_dict = {
-            "bn": bn,
+            **globals(),
```
[the remark this answers is not in the document]
Comment (namanjain) [reply tests/test_flow/test_persistence_random.py:203]: This is directly related to one of your [previous comment](https://github.com/square/bionic/pull/299#discussion_r530565556). I didn't know that functions can have a `None` module. But I uncovered that from this test. If we don't pass it `**globals()` here, then the function that we define here does not have a module associated with it.

If you think adding all globals is excessive, I can use something like `import_code` helper method for this instead. Wdyt?

File: tests/test_flow/test_persistence.py [hunk 8]
```diff
@@ -747,6 +748,139 @@ def y():  # noqa: F811
     assert f_call_counter.times_called() == 0
 
 
+def test_versioning_auto_with_local_refs(builder, make_counter):
```
[the remark this answers is not in the document]
Comment (namanjain) [reply tests/test_flow/test_persistence.py:751]: Sure, I can add some tests for assisted versioning as well.

**Comment by namanjain** [comment]
> It's really exciting to see this coming together! A few thoughts:
> 
> 1. Would it make sense to just combine this with your other PR?

Yeah, I can merge this with my other PR.

> 2. We still need to add release notes and update the documentation, right?

Right, that's on my list of things to do (you can find that in the PR description). I felt like this PR was already too big. But I can add documentation and release notes with this change if you prefer.

> 3. Have you noticed any performance impact from these changes? Do the tests take longer?

I haven't noticed any difference for the baseline tests. Even for the entire test suite, it took maybe a minute more with the new change, but I ran it only once locally and there is always a lot of variance.

> 4. These changes affect assisted versioning too, right? Do we need to change how versioning errors are reported?

What kind of versioning errors do you have in mind?

> 5. The tests here are good, but shouldn't we also have some tests on more complex code? Your hashing algorithm is quite complex and stateful, so it feels like it has a lot of state space that isn't explored by these relatively small tests.
> 
> That last point is a hard one. Some dumb ideas:
> 
> 1. Using the `ml_workflow_cli` example. (It's not actually that complex, though -- and we wouldn't want to make our tutorial code more complex for the sake of this test. But it would actually be great for our examples to be run as part of our test suite.)
> 2. Generating random code. (You'll only be able to test the constructs that you think to generate, but at least you'll be able to test how the different combinations interact.)
> 3. Importing (or vendoring) some open source library and trying to hash some functions in it; then monkey-patching some of the other functions or constants. (Kind of hacky but saves you from coming up with your own code.)

Yeah, I definitely want to add more complex tests. (That's another one of my things to do in the PR description and I wanted to get both of your thoughts on it.) Let's talk about this one separately.

I will also decline this PR for now and move the changes to #232. I don't know if we can still continue the conversation on this PR, but if not, you can start a new thread with a link to the older thread (from this PR) there (the older PR).

File: bionic/code_hasher.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:201]: I think it can also be a free variable.

File: bionic/datatypes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/datatypes.py:308]: Yes, they are ingested through `@version`, but our API doesn't need to be 1:1 with our internal data model. And I would argue that neither `suppress_bytecode_warnings` nor `ignore_bytecode` are really part of the version; they're per-entity configuration options that affect how we interpret the version. For example, if someone asked "what is the version of this entity", I would expect the answer to include the major and minor numbers (maybe as a human-readable string), but not the `ignore_bytecode` settings. Similarly, it doesn't line up with what's described in the docstring of `CodeVersion`.

Of course, that's not to say that we can't stick these on `CodeVersion` anyway to make things simpler. But since this class gets serialized as part of our cache schema, it will be annoying to change later, so I just want to make sure we're thoughtful about why we're doing this.

[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/datatypes.py:308]: After thinking about this some more, I've partly come around to your point of view here: maybe it does make sense for these fields to be here. Here's my current thinking:

1. Having things that look like "instructions" on a version object seems weird to me. I feel like a version should be more like a piece of data that gets interpreted rather than a set of instructions. However, if we think of `ignore_bytecode` as more like `includes_bytecode` (changing from an instruction to a description) then it makes more sense to me. This is similar to the difference between `@persist` and `@serializable`.

2. I don't have a similar rationale for `suppress_bytecode_warnings` but you can't win them all. :)

File: bionic/decorators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/decorators.py:90]: Ah, fair point. I see the other fields have converters because our internal code also creates CodeVersions. My main concerns is that we have to document the default behavior in the `@version` docstring, so it's a bit weird to have the actual transformation taking place elsewhere.

File: bionic/datatypes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/datatypes.py:308]: I guess another possibility would be to have a `CodeVersioningPolicy` object that includes a `CodeVersion` along with all the additional settings.

File: bionic/code_hasher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/code_hasher.py:72]: Oh, you're right, just requiring it to be non-`None` won't work. Let me rephrase my worry:

`code_context` is sometimes `None` and sometimes not, and there are some code paths that assume that it's not `None`. So how are we guaranteeing that it will always be present at the time that it's needed? I'm sure there's a reason it works, but it's not immediately apparent; I'd have to trace through all the variations to be convinced. Ideally the code would be structured so that it's obviously impossible for it to be missing when we need it, but we could also just have a comment that explains why this is safe.

(BTW, if we call `CodeHasher.hash` on a code object, will it work?)

File: tests/test_flow/test_persistence_random.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply tests/test_flow/test_persistence_random.py:203]: I think I'm missing two things here:

1. If we just want this function to have a `__module__`, why do we need to include all of `globals()`? Is there a particular global variable that determines what the module is?

2. Why do we choose not to scan functions that have empty modules?

**Comment by jqmp** [comment]
> I felt like this PR was already too big. But I can add documentation and release notes with this change if you prefer.

My view here is that we should try to always keep master in a "good" state, which includes correct documentation and up-to-date release notes.

More pragmatically, putting the documentation and release notes in the same PR makes it easier for the reviewer to understand what the PR is doing, and also increases the chances of catching any discrepancies between the docs and the implementation. There's also the chance that writing the docs will surface some design issues that make you want to go back and change your implementation before submitting the PR.

(So basically this is just a special case of my view that a PR should be self-contained.)

File: bionic/datatypes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/datatypes.py:308]: I think I'll do both, change `ignore_bytecode` to `includes_bytecode` and create a `CodeVersioningPolicy` object. I still think `includes_bytecode` should be a part of code version, but it would be good to have a policy object to contain other fields.

File: bionic/code_hasher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:72]: Oh I see. We really just need `code_context` to be present for code objects. We will get code objects only from a coroutine in the current implementation, and that will always create a code object for you. I can add an assert in `_ingest_code` method that `code_context` is not null, since that is the only place where we actually use the `code_context`. Wdyt?

To answer your question, `CodeHasher.hash` won't work on a code object. That would throw an error.

File: tests/test_flow/test_persistence_random.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply tests/test_flow/test_persistence_random.py:203]: 1. Right, passing `globals()` also passes some variable that defines the module for the exec code.

2. Well, we decided to not hash any variables that are either in `Bionic`, or is an internal module (anything that lives in `sysconfig_paths()`, which is what our reload functionality does too).
I don't have a good reason to include or not include coroutines without modules. There also isn't enough information out there that tells us when `__module__` object isn't available. So I decided to ignore these objects since they can have unintended side effects. But I can be easily convinced otherwise.

[the remark this answers is not in the document]
Comment (jqmp) [reply tests/test_flow/test_persistence_random.py:203]: 1. So could we just pass in that particular variable?

2. What side effects did you have in mind? I don't have a good sense either about whether to include these or not, but I would probably err on the side of hashing them. (My intuition is that these functions are more likely to be user functions than library functions, but I could be wrong.) Either way, let's leave a comment with our rationale.

**Comment by jqmp** [comment]
> What kind of versioning errors do you have in mind?

Sorry, I missed this question earlier. I was thinking of the `CodeVersioningError` that we raise when the bytecode doesn't match. Right now we say "the code function that outputs ... was changed ... but the function's version number was not". Should we rephrase this?

File: bionic/code_hasher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/code_hasher.py:72]: Is there anything sensible we could do if we got passed a code object? Or is throwing an exception our only option?

If the former, then a `DEFAULT_CONTEXT` might make sense. If not, then I agree putting an assert there makes sense. (Plus a comment explaining why it can't happen, of course.)

File: tests/test_flow/test_persistence_random.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply tests/test_flow/test_persistence_random.py:203]: On second thought, we only check module to see if it's a Bionic variable, and any Bionic variable should have that variable. So I'll hash variables if thy don't have a variable name.

File: bionic/code_hasher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (namanjain) [reply bionic/code_hasher.py:72]: We can hash only the code object and skip the references. I don't think we can find references in that case. I'll add a `DEFAULT_CONTEXT` and add a field called `can_find_references` in `CodeContext`.
